### PR TITLE
#1977 - Request Offering Change: Student Application Tracker - Part 2

### PR DESCRIPTION
**Pull request by sh16011993**
### As a part of this PR, the following are completed:

**1.** On app tracker, in completed status:
- added a warning card with button "Review changes" when status is in progress with student.
- added a warning card showing the request is pending SABC approval if the student approves the change request.
- added a student declined card if the student declines the change request. 
- added a ministry approved card if the ministry approves the change request. 
- added a ministry declined card if the ministry declines the change request.
 
 Note: All the statuses are based on the most recent submitted request for change (Same logic from student appeal).

**2.** Added logic to display the offering change requests in the "assessment" table.

**3.** Modified the logic to show amber color in the application tracker when there is atleast one of the following in progress with the student application - appeal, exception, offering request or offering change request.

**Screenshots:**

Application Tracker View:

- Application Offering Change Request In Progress with the Student

# image
                      
 
- Application Offering Change Request in Progress with the Ministry

# image

 
- Application Offering Change Request Declined by the Student

# image

 
- Application Offering Change Request Declined by the Ministry

image

 
- Application Offering Change Request Approved by the Ministry

# image
 
**Requests View:**

# image

**History View:**

# image

**To be done in the next (final) PR:**

- e2e tests

### CODE REVIEW

File: sources/packages/backend/apps/api/src/app.aest.module.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/app.aest.module.ts:170]: do we need to import it into aest module?

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:547]: this check is already happening i line 537 right?

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:549]: I believe, I asked you in our call. were these details not be able to retrieve from line 526 service??

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:323]: do we need `orderby` if its `findOne`?

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:573]: IMO, I don't think you need to specify `most recent`. Because, status are always most recent

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:631]: same here, https://github.com/bcgov/SIMS/pull/2124/files#r1270092485

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts:145]: pls update the method comment too

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:331]: pls review the commnet

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:348]: just double checking, do we need `relations` just for the sake of `where`. I may be wrong, but I remember a comment in a PR, stating something like that, can you double-check?

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts:156]: Not a blocker, For the appeal, we are doing the dto transformation in the service (which will be specific for this endpoint), then you can do something like this 
image

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts:166]: I believe, we don't use the object type name in the variable name, unless it's needed

File: sources/packages/backend/libs/sims-db/src/entities/assessment-trigger.type.ts
Comment (ann-aot) [inline sources/packages/backend/libs/sims-db/src/entities/assessment-trigger.type.ts:40]: `Triggered when the Institution creates an offering change request for the student which required student and SABC approval that will result in the assessment ...`?

File: sources/packages/web/src/components/students/applicationTracker/Completed.vue
Comment (dheepak-aot) [inline sources/packages/web/src/components/students/applicationTracker/Completed.vue:56]: It is not a must, but I would recommend to have the application tracker screenshots for following remaining scenarios as well.

 - added a warning card showing the request is pending SABC approval if the student approves the change request.
 - added a student declined card if the student declines the change request.
 - added a ministry approved card if the ministry approves the change request.
 - added a ministry declined card if the ministry declines the change request.

File: sources/packages/web/src/components/students/applicationTracker/ApplicationProgressBar.vue
Comment (ann-aot) [inline sources/packages/web/src/components/students/applicationTracker/ApplicationProgressBar.vue:127]: @dheepak-aot  @guru-aot do you why ` if (
        props.applicationStatus === ApplicationStatus.Completed &&
        applicationProgressDetails.value.appealStatus !==
          StudentAppealStatus.Pending
      )` was done instead of `(statusIconDetails.value.statusType === "success")` before, just wanted to check, if there was a reason

File: sources/packages/web/src/components/students/applicationTracker/ApplicationProgressBar.vue
Comment (ann-aot) [inline sources/packages/web/src/components/students/applicationTracker/ApplicationProgressBar.vue:129]: there is no need of `else` I believe

File: sources/packages/web/src/components/students/applicationTracker/Completed.vue
Comment (ann-aot) [inline sources/packages/web/src/components/students/applicationTracker/Completed.vue:90]: pls use keyboard single quote for `You’ve`

File: sources/packages/web/src/services/StudentAssessmentsService.ts
Comment (ann-aot) [inline sources/packages/web/src/services/StudentAssessmentsService.ts:23]: do we need `in progress ` ?? because it is already mentioned in line 22

File: sources/packages/web/src/types/contracts/AssessmentTrigger.ts
Comment (ann-aot) [inline sources/packages/web/src/types/contracts/AssessmentTrigger.ts:40]: https://github.com/bcgov/SIMS/pull/2124/files#r1270101378

File: sources/packages/web/src/types/contracts/StatusChip.ts
Comment (ann-aot) [inline sources/packages/web/src/types/contracts/StatusChip.ts:9]: It will be better, if you add the intention of the enum as comment

**Review by ann-aot — COMMENTED** [review]
Added some comments

File: sources/packages/web/src/components/students/applicationTracker/ApplicationProgressBar.vue
Comment (dheepak-aot) [reply sources/packages/web/src/components/students/applicationTracker/ApplicationProgressBar.vue:127]: I am not sure of the exact reason @ann-aot . When I had a look at the first PR of #1683 I see the status icon was missing on various scenarios on completed status. But the upcoming PRs added the status icon back on these scenarios I guess.  But not 100% sure.

###PR 1
image

File: sources/packages/backend/apps/api/src/app.aest.module.ts
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/app.aest.module.ts:170]: You're right. we don't need them in AEST module.

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:323]: @ann-aot I believe it is to guarantee the return of most recent application offering change.

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:548]: @ann-aot @andrewsignori-aot @guru-aot I have a question here. 

Do we need the `mostRecentApplicationOfferingChangeRequestStatus` and `appealStatus` returned as a part of `application/in-progress` API?

I believe they are required only for `application/completed` API? as these values are used only by Completed.vue right?

Let me know if I am missing something

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts:156]: I agree. We can move the transformation to controller service.

Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts:156]: I would recommend to avoid technical terms like "Array" in variable name.

File: sources/packages/web/src/components/generic/StatusChipRequestedAssessment.vue
Comment (dheepak-aot) [inline sources/packages/web/src/components/generic/StatusChipRequestedAssessment.vue:28]: I am not able to follow why did we create an additional mapping for label here.

File: sources/packages/web/src/components/students/applicationTracker/Completed.vue
Comment (dheepak-aot) [inline sources/packages/web/src/components/students/applicationTracker/Completed.vue:57]: I believe after scholastic standing, the order of card arrangements must be based on status.

If we see the the in-progress component we arrange all the denial cards and then the waiting cards. 

I believe here, we must put all the in-progress status  cards(after scholastic standing cards), and then follow the same existing order for other status.

@ann-aot @andrewsignori-aot @guru-aot  thoughts?

**Comment by dheepak-aot** [comment]
Nice work @sh16011993 . Please have a look at the comments.

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (guru-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:548]: I believe, this is not required here. You are right @dheepak-aot

Comment (ann-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:548]: you are correct @dheepak-aot. But just wondering if this is `application/${applicationId}/in-progress` endpoint?

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:548]: I am not sure if I am following this, but this method returns information for the "application progress"  not the application "in progress" status. This method can be called at any time (any application status) to provide overall statuses for the application, mainly used for the "application tracker/progress bar", hence we need all the information including appeals and offering request status.
Please let me know If I am missing something.

Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:548]: Both DB calls can happen in parallel now.

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (sh16011993) [reply sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:348]: Thank You @ann-aot. I removed the `relations`. It is not needed in this scenario.

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:573]: I agree with @ann-aot. We can follow the same from appeals.

File: sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/assessment/assessment.students.controller.ts:156]: I also agree. Please notice that the comments have two points, please do both.
1. Move the transformation to the controller service.
2. Execute the calls in parallel.

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:327]: As per the comment from @dheepak-aot (https://github.com/bcgov/SIMS/pull/2124/files#r1270971907) making the method more generic would be better, which would also include returning the object instead of the specific property, IMO.

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:348]: We do not need relations for the `where`.

File: sources/packages/web/src/components/students/applicationTracker/Completed.vue
Comment (sh16011993) [reply sources/packages/web/src/components/students/applicationTracker/Completed.vue:56]: Thanks @dheepak-aot. Added the screenshots.

File: sources/packages/web/src/components/generic/StatusChipRequestedAssessment.vue
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/generic/StatusChipRequestedAssessment.vue:28]: If I am not wrong, only for Students, there was a UI request to change some of them to be displayed differently, so, besides mapping only the statuses to colors it was also needed to have some label "translated" again. I am not defending the idea, just trying to recap what was requested from the UI perspective.
That said, it was only meant to be displayed like that for Students, so I am not sure if the generic component should be the one dealing with it.

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (sh16011993) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:548]: Yes @andrewsignori-aot , you are right.

File: sources/packages/backend/apps/api/src/app.aest.module.ts
Comment (sh16011993) [reply sources/packages/backend/apps/api/src/app.aest.module.ts:170]: Without adding it here, I get an error stating that Nest cannot resolve dependency related to `ApplicationOfferingChangeRequestService`. Hence, keeping it here as is. Please correct me if I am wrong.

Comment (ann-aot) [reply sources/packages/backend/apps/api/src/app.aest.module.ts:170]: It was not throwing an error, When I first reviewed it. Now since you have added the service to `AssessmentControllerService`,  removing it will throw an error

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts]: I am sorry guys. I totally lost track of difference between the APIs `in-progress` and `progress-details`. (I created the `progress-details` API I believe 😊) I can recollect now the progress details was to support the progress of the bar.

File: sources/packages/web/src/components/students/applicationTracker/Completed.vue
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/students/applicationTracker/Completed.vue:57]: I see the point but would it not possibly mess with the order that they are displayed?
For instance, for "student appeals" and "offering change request" we can have a complete card and also an in-progress card displayed at the same time. In this way, they are displayed as a group. Does it make sense?

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:318]: mm, Once an application offering request is declined or approved, the institution can again request another application offering request, right? so, that means, there can be multiple rows that needed to be displayed in the tracker and banners right? correct me if I am wrong.

File: sources/packages/web/src/components/students/applicationTracker/Completed.vue
Comment (ann-aot) [reply sources/packages/web/src/components/students/applicationTracker/Completed.vue:57]: I actually, didn't get the qn

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts
Comment (sh16011993) [reply sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:318]: When there is a re-request, from my understanding it will replace the existing row of the previous request in the banner. There is only one row for each of the appeal and application offering change request in the banner. Please correct me if I am wrong.

Comment (ann-aot) [reply sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:318]: or do we only care about the latest request?

File: sources/packages/web/src/components/students/applicationTracker/Completed.vue
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/students/applicationTracker/Completed.vue:57]: As the Team agreed, let's have them organized as below:
- Anything In progress
- Anything declined
- Anything trigger type based
- Disbursement/COE banners

**Review by andrewsignori-aot — COMMENTED** [review]
Good work, thanks for doing the changes. Once the reorganization on `applicationTracker/Completed.vue` is done I will approve it.

File: sources/packages/web/src/components/students/applicationTracker/Completed.vue
Comment (dheepak-aot) [inline sources/packages/web/src/components/students/applicationTracker/Completed.vue:102]: As discussed, the error card content needs to be updated as per figma.

**Review by guru-aot — APPROVED** [review]
NIce work. Thanks for doing the changes

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes  and the ordering of cards👍

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for re-ordering the cards, looks good 👍

**Pull request merged**
